### PR TITLE
Add privilege checker and bind it to session

### PR DESCRIPTION
**Pull request by shenli**

### CODE REVIEW

File: privilege/privileges/privileges.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+	if !ok {
+		return false, nil
+	}
+	_, ok = tblp[tbl.Name.O]
+	return ok, nil
+}
+
+func (p *PrivilegeCheck) loadGlobalPrivileges(ctx context.Context) error {
+	sql := fmt.Sprintf(`SELECT * FROM %s.%s WHERE User="%s" AND (Host="%s" OR Host="%%");`, mysql.SystemDB, mysql.UserTable, p.username, p.host)
+	rs, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, sql)
```
Comment (siddontang) [inline privilege/privileges/privileges.go]: The struct name is RestrictedSQLExecutor, so I think using Exec as the function name may be simple and is already clear to show that we will exec a restricted SQL.

File: privilege/privileges/privileges.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+	}
+	for {
+		row, err := rs.Next()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if row == nil {
+			break
+		}
+		for i := 3; i < len(fs); i++ {
```
Comment (siddontang) [inline privilege/privileges/privileges.go]: why start from 3 here? a magic number.

File: privilege/privileges/privileges.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+func (p *PrivilegeCheck) loadTableScopePrivileges(ctx context.Context) error {
+	sql := fmt.Sprintf(`SELECT * FROM %s.%s WHERE User="%s" AND (Host="%s" OR Host="%%");`, mysql.SystemDB, mysql.TablePrivTable, p.username, p.host)
+	rs, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer rs.Close()
+	ps := make(map[string]map[string]map[mysql.PrivilegeType]bool)
```
Comment (siddontang) [inline privilege/privileges/privileges.go]: em...... seem too nested, why not define a struct or a type?

File: privilege/privileges/privileges.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+		}
+	}
+	p.privs.TablePrivs = ps
+	return nil
+}
+
+func (p *PrivilegeCheck) loadPrivileges(ctx context.Context) error {
+	user := variable.GetSessionVars(ctx).User
+	strs := strings.Split(user, "@")
+	p.username, p.host = strs[0], strs[1]
```
Comment (siddontang) [inline privilege/privileges/privileges.go]: no need to check strs' item number?

File: privilege/privilege.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+// Checker is the interface for check privileges.
+type Checker interface {
+	CheckDBPrivilege(ctx context.Context, db *model.DBInfo, privilege mysql.PrivilegeType) (bool, error)
+	CheckTablePrivilege(ctx context.Context, db *model.DBInfo, tbl *model.TableInfo, privilege mysql.PrivilegeType) (bool, error)
+}
+
+const key keyType = 0
+
+// BindPrivilegeChecker binds domain to context.
```
Comment (qiuyesuifeng) [inline privilege/privilege.go]: binds Checker to context?

File: privilege/privilege.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+func BindPrivilegeChecker(ctx context.Context, pc Checker) {
+	ctx.SetValue(key, pc)
+}
+
+// GetPrivilegeChecker gets domain from context.
```
Comment (qiuyesuifeng) [inline privilege/privilege.go]: gets Checker?

File: privilege/privileges/privileges.go [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+	if !ok {
+		return false, nil
+	}
+	_, ok = dbp[privilege]
+	return ok, nil
+}
+
+// CheckTablePrivilege implements PrivilegeChecker.CheckTablePrivilege interface.
+func (p *PrivilegeCheck) CheckTablePrivilege(ctx context.Context, db *model.DBInfo, tbl *model.TableInfo, privilege mysql.PrivilegeType) (bool, error) {
+	if p.privs == nil {
```
Comment (qiuyesuifeng) [inline privilege/privileges/privileges.go]: Any need to use lock to protect `p.privs == nil`?

File: privilege/privileges/privileges.go [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+		return true, nil
+	}
+	// Check db scope privileges.
+	dbp, ok := p.privs.DBPrivs[db.Name.O]
+	if ok {
+		_, ok = dbp[privilege]
+		if ok {
+			return true, nil
+		}
+	}
```
Comment (qiuyesuifeng) [inline privilege/privileges/privileges.go]: Seems can use CheckDBPrivilege above?

File: privilege/privileges/privileges.go [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+	tblp, ok := p.privs.TablePrivs[db.Name.O]
+	if !ok {
+		return false, nil
+	}
+	_, ok = tblp[tbl.Name.O]
+	return ok, nil
+}
+
+func (p *PrivilegeCheck) loadGlobalPrivileges(ctx context.Context) error {
+	sql := fmt.Sprintf(`SELECT * FROM %s.%s WHERE User="%s" AND (Host="%s" OR Host="%%");`, mysql.SystemDB, mysql.UserTable, p.username, p.host)
```
Comment (qiuyesuifeng) [inline privilege/privileges/privileges.go]: A little confused that if use (Host="%s" OR Host="%%") , Host="%s" will be useless?

File: privilege/privileges/privileges.go (hunk 1 above)
Comment (shenli) [inline privilege/privileges/privileges.go]: session implements RestrictedSQLExecutor. But it already have a function named Execute. So Exec may be confused with Execute.

File: privilege/privileges/privileges.go (hunk 2 above)
Comment (c4pt0r) [inline privilege/privileges/privileges.go]: +1

Comment (shenli) [inline privilege/privileges/privileges.go]: I will add some comments there.

File: privilege/privileges/privileges.go (hunk 4 above)
Comment (shenli) [inline privilege/privileges/privileges.go]: User name comes from parser or mysql client. It is already checked.

File: privilege/privileges/privileges.go (hunk 9 above)
Comment (shenli) [inline privilege/privileges/privileges.go]: Host="%%" means all host is ok. User has privileges granted to user@host and user@%.

File: privilege/privileges/privileges.go (hunk 7 above)
Comment (shenli) [inline privilege/privileges/privileges.go]: Each user has a Checker in a session. This will not be called concurrently.

**Comment by shenli** [comment]
PTAL

File: privilege/privilege.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+const key keyType = 0
+
+// BindPrivilegeChecker binds Checker to context.
+func BindPrivilegeChecker(ctx context.Context, pc Checker) {
+	ctx.SetValue(key, pc)
+}
+
+// GetPrivilegeChecker gets Checker from context.
+func GetPrivilegeChecker(ctx context.Context) Checker {
+	v, ok := ctx.Value(key).(Checker)
```
Comment (c4pt0r) [inline privilege/privilege.go]: if v, ok := ctx.Value(key).(Checker); ok {
    return v
}
return nil

Comment (shenli) [inline privilege/privilege.go]: PTAL

File: privilege/privilege.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+const key keyType = 0
+
+// BindPrivilegeChecker binds Checker to context.
+func BindPrivilegeChecker(ctx context.Context, pc Checker) {
+	ctx.SetValue(key, pc)
+}
+
+// GetPrivilegeChecker gets Checker from context.
+func GetPrivilegeChecker(ctx context.Context) Checker {
```
Comment (siddontang) [inline privilege/privilege.go]: no place to use this function?

File: privilege/privileges/privileges.go [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+	}
+	// Check db scope privileges.
+	dbp, ok := p.privs.DBPrivs[db.Name.O]
+	if ok {
+		ok = dbp.contain(privilege)
+		if ok {
+			return true, nil
+		}
+	}
+	if tbl == nil {
```
Comment (qiuyesuifeng) [inline privilege/privileges/privileges.go]: How about move this check to the first of function?

Comment (qiuyesuifeng) [inline privilege/privileges/privileges.go]: Perhaps not much different, ignore it.

File: privilege/privileges/privileges.go [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+		row, err := rs.Next()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if row == nil {
+			break
+		}
+		// mysql.User table privilege columns start from index 3.
+		// See: booststrap.go CreateUserTable
+		for i := 3; i < len(fs); i++ {
```
Comment (qiuyesuifeng) [inline privilege/privileges/privileges.go]: Any need to use a const variable instead of 3?

File: privilege/privileges/privileges.go [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+			d := row.Data[i]
+			ed, ok := d.(mysql.Enum)
+			if !ok {
+				return fmt.Errorf("Privilege should be mysql.Enum: %v(%T)", d, d)
```
Comment (qiuyesuifeng) [inline privilege/privileges/privileges.go]: errors.Errorf

File: privilege/privilege.go (hunk 11 above)
Comment (shenli) [inline privilege/privilege.go]: It will be used in the next pr.

**Comment by shenli** [comment]
PTAL @qiuyesuifeng

File: privilege/privileges/privileges.go [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if row == nil {
+			break
+		}
+		// DB
+		db, ok := row.Data[1].(string)
+		if !ok {
+			panic("This should be never happened!")
```
Comment (siddontang) [inline privilege/privileges/privileges.go]: I prefer error instead of panic.

Comment (shenli) [inline privilege/privileges/privileges.go]: error means expected excption. But if ok is false, there must be something wrong. And it is a fatal error.

Comment (siddontang) [inline privilege/privileges/privileges.go]: It is dangerous to use panic, we may forget releasing resource, or others, and even more, if we don't use defer outside, whole program will crash.

**Comment by shenli** [comment]
PTAL @siddontang

**Comment by siddontang** [comment]
LGTM

**Pull request merged**